File: python/codeql/full_ssrf/benign.py
import requests

from flask import Flask, request


app = Flask(__name__)
ALLOWED_URLS = {
    "status": "https://status.example.invalid/health",
    "docs": "https://docs.example.invalid/",
}


@app.get("/fetch-safe")
def fetch_safe():
    url = ALLOWED_URLS.get(request.args.get("target", "status"), ALLOWED_URLS["status"])
    return requests.get(url, timeout=5).text

File: javascript/codeql/request_forgery/benign.js
const axios = require("axios");
const express = require("express");

const app = express();
const targets = new Map([["status", "https://status.example.invalid/health"], ["docs", "https://docs.example.invalid/"]]);

app.get("/fetch-safe", async (req, res) => {
  const url = targets.get(String(req.query.target || "status")) || targets.get("status");
  const response = await axios.get(url);
  res.send(response.data);
});
